motionplan/constraint.go

```diff
@@ -493,3 +493,154 @@ func NewOctreeCollisionConstraint(octree *pointcloud.BasicOctree, threshold int,
+	Frame1 string
+	Frame2 string
+}
+
+// CollisionSpecification is used to selectively apply obstacle avoidance to specific parts of the robot
+type CollisionSpecification struct {
+	// Pairs of frame which should be allowed to collide with one another
+	Allows []*CollisionSpecification_AllowedFrameCollisions
+}
+
+// Constraints is a struct to store the constraints imposed upon a robot
+// It serves as a convenenient RDK wrapper for the protobuf object
+type Constraints struct {
+	LinearConstraint       []*LinearConstraint
+	OrientationConstraint  []*OrientationConstraint
+	CollisionSpecification []*CollisionSpecification
+}
+
+func NewEmptyConstraints() *Constraints {
+	return &Constraints{
+		LinearConstraint:       make([]*LinearConstraint, 0),
+		OrientationConstraint:  make([]*OrientationConstraint, 0),
+		CollisionSpecification: make([]*CollisionSpecification, 0),
+	}
+}
+
+func NewConstraints(linConstraints []*LinearConstraint, orientConstraints []*OrientationConstraint, collSpecifications []*CollisionSpecification) *Constraints {
+	return &Constraints{
+		LinearConstraint:       linConstraints,
+		OrientationConstraint:  orientConstraints,
+		CollisionSpecification: collSpecifications,
+	}
+}
+
+func ConstraintsFromProtobuf(pbConstraint *motionpb.Constraints) *Constraints {
+	// iterate through all motionpb.LinearConstraint and convert to RDK form
+	linConstraintFromProto := func(linConstraints []*motionpb.LinearConstraint) []*LinearConstraint {
+		toRet := make([]*LinearConstraint, 0)
+		for _, linConstraint := range linConstraints {
+			toRet = append(toRet, &LinearConstraint{
+				LineToleranceMm:          linConstraint.LineToleranceMm,
+				OrientationToleranceDegs: linConstraint.OrientationToleranceDegs,
+			})
+		}
+		return toRet
+	}
+
+	// iterate through all motionpb.OrientationConstraint and convert to RDK form
+	orientConstraintFromProto := func(orientConstraints []*motionpb.OrientationConstraint) []*OrientationConstraint {
+		toRet := make([]*OrientationConstraint, 0)
+		for _, orientConstraint := range orientConstraints {
+			toRet = append(toRet, &OrientationConstraint{
+				OrientationToleranceDegs: orientConstraint.OrientationToleranceDegs,
+			})
+		}
+		return toRet
+	}
+
+	// iterate through all motionpb.CollisionSpecification and convert to RDK form
+	collSpecFromProto := func(collSpecs []*motionpb.CollisionSpecification) []*CollisionSpecification {
+		toRet := make([]*CollisionSpecification, 0)
+		for _, collSpec := range collSpecs {
+			allowedFrameCollisions := make([]*CollisionSpecification_AllowedFrameCollisions, 0)
+			for _, collSpecAllowedFrame := range collSpec.Allows {
+				allowedFrameCollisions = append(allowedFrameCollisions, &CollisionSpecification_AllowedFrameCollisions{
+					Frame1: collSpecAllowedFrame.Frame1,
+					Frame2: collSpecAllowedFrame.Frame2,
+				})
+			}
+		}
+		return toRet
+	}
+
+	return NewConstraints(linConstraintFromProto(pbConstraint.LinearConstraint),
```

Comment:
Another go style thing: if you want to break up arguments over multiple lines we're either going to put them all on different lines or all on the same line.  So in this case
```
NewConstraints(
   linear,
   orientation,
   collSpec,
)
```